File: Searching/JavaCodes/Leetcode1011.java
class Solution {
    public static boolean canITransport(int[] weights,int days,int capacity){
        int shipWeight=0;
        int tempDays=1;

        for (int i = 0; i < weights.length; i++) {
            if(shipWeight+weights[i]<=capacity){
                shipWeight+=weights[i];
            }
            else{
                tempDays++;
                shipWeight=weights[i];
            }
        }
        return tempDays<=days;
    }
    public static int shipWithinDays(int[] weights, int days) {
        int max=0,sum=0;

        for (int i = 0; i < weights.length; i++) {
            max=Math.max(max,weights[i]);
            sum+=weights[i];
        }
        
        int start=max,end=sum;
        
        while(start<=end){
            int mid=start+(end-start)/2;
            if(canITransport(weights,days,mid)) {
                end = mid - 1;
            }
            else{
                start=mid+1;
            }
        }
        return start;
    }
}

File: Searching/PythonCodes/Leetcode1011.py
class Solution(object):
    def canITransport(self, weights, days, capacity):
        ship_weight = 0
        temp_days = 1

        for weight in weights:
            if ship_weight + weight <= capacity:
                ship_weight += weight
            else:
                temp_days += 1
                ship_weight = weight

        return temp_days <= days

    def shipWithinDays(self, weights, days):
        max_weight = max(weights)
        total_weight = sum(weights)

        start, end = max_weight, total_weight

        while start <= end:
            mid = start + (end - start) // 2
            if self.canITransport(weights, days, mid):
                end = mid - 1
            else:
                start = mid + 1

        return start
